ouroboros-consensus/src-docs/Test/Tutorial2.lhs

```diff
@@ -0,0 +1,691 @@
+% Example: A More Complex Example Protocol Involving Epochs
+
+Introduction
+============
+
+This tutorial builds on the example in `Tutorial.lhs` by specifying
+a block, ledger, and protocol for a blockchain that has a notion
+of epoch built into it - serving as an example of a case where
+the state of the ledger is used as part of the machinery
+of consensus.  It is highly recommended that the reader have
+read `Tutorial.lhs` before this if they are not already
+familiar with concepts such as `ConsensusProtocol` and
+`LedgerState`.
+
+Much like the previous example of `BlockC` this blockchain
+(with block type `BlockD`) models a single number resulting
+from a series of _increment_ and _decrement_ transactions in
+the block bodies.  However, the slots are now divided into
+**epochs** each of which has a fixed number of slots.
+
+Further, the chain is set up such that there are 20 _nodes_
+labelled with a `NodeId` from 0 to 19 participating in the
+computation.
+
+At the beginning of an epoch, the current value of the
+ledger - that is, the sum of the _increment_ transactions minus
+the sum of the _decrement_ transactions is snapshotted as
+part of the ledger.
+
+During the epoch, the snapshot *from two epochs ago* determines
+which subset of a set of
+20 nodes is allowed to lead slots.  If the snapshot is even,
+then slots contained in that epoch follow a round-robin leadership schedule
+among nodes 0 through 9 inclusive.  Similarly, if the snapshot is odd then
+the slots of that epoch will be follow a round-robin leadership schedule
+from the set of nodes from 10 to 19.
+
+Though it is difficult to imagine a real system doing this,
+it is a simple example of a case where the value computed
+by the blockchain (the `LedgerState`) is relevant - through the
+leader schedule - to the behavior of consensus.  One can perhaps view
+it as a very simplified analog to stake distribution.
+
+Setup
+-----
+
+As before, we require a few language extensions:
+
+> {-# OPTIONS_GHC -Wno-unused-top-binds   #-}
+> {-# LANGUAGE TypeFamilies               #-}
+> {-# LANGUAGE DerivingVia                #-}
+> {-# LANGUAGE DataKinds                  #-}
+> {-# LANGUAGE DeriveGeneric              #-}
+> {-# LANGUAGE FlexibleInstances          #-}
+> {-# LANGUAGE MultiParamTypeClasses      #-}
+> {-# LANGUAGE GeneralizedNewtypeDeriving #-}
+> {-# LANGUAGE DeriveAnyClass #-}
+> {-# LANGUAGE StandaloneDeriving         #-}
+
+> module Test.Tutorial2() where
```

Comment:
suggestion: rename to something more self-describing like `Ouroboros.Consensus.Tutorial.WithEpoch`